File: SEM 6/Computer Networks/chapter1/19.c
//  Write a function reverse(s) that reverses the character string s. Use it to
//  write a program that reverses its input a line at a time.


#include <stdio.h>
#include <string.h>

#define MAX_LINE 1000  // Maximum line length

// Function to reverse a string in place
void reverse(char s[]) {
    int len = strlen(s);
    int i, j;
    char temp;
    
    for (i = 0, j = len - 1; i < j; i++, j--) {
        temp = s[i];
        s[i] = s[j];
        s[j] = temp;
    }
}

//  abcd --> dbca


// Function to read a line from input
int get_line(char s[], int limit) {
    int c, i;
    for (i = 0; i < limit - 1 && (c = getchar()) != EOF && c != '\n'; i++) {
        s[i] = c;
    }
    if (c == '\n') {
        s[i++] = c;  // Retain newline character
    }
    s[i] = '\0';
    return i;
}

// 1234567






int main() {
    char line[MAX_LINE];

    // printf("Enter text (Ctrl+D to stop on Linux/macOS, Ctrl+Z on Windows):\n");

    while (get_line(line, MAX_LINE) > 0) {
        reverse(line);
        printf("%s", line);  // Print reversed line
    }

    return 0;
}

File: SEM 6/Computer Networks/chapter1/19.lua
-- Define the maximum line length
local MAX_LINE = 1000

-- Function to reverse a string in place (Lua strings are immutable, so we use a table)
local function reverse(s)
    local chars = {}
    for i = 1, #s do
        chars[i] = s:sub(i, i)  -- Split the string into individual characters
    end

    -- Reverse the characters
    for i = 1, math.floor(#chars / 2) do
        local temp = chars[i]
        chars[i] = chars[#chars - i + 1]
        chars[#chars - i + 1] = temp
    end

    -- Rebuild the reversed string
    return table.concat(chars)
end

-- Function to read a line from input
local function get_line(limit)
    local s = io.read("*line")  -- Read a line from standard input
    if s then
        if #s > limit - 1 then
            s = s:sub(1, limit - 1)  -- Truncate if it exceeds the limit
        end
        return s .. "\n"  -- Append newline character
    else
        return ""  -- Return empty string on EOF
    end
end

-- Example usage
print("Enter a line of text:")
local line = get_line(MAX_LINE)
print("You entered:\n", line)

local reversed_line = reverse(line)
print("Reversed line:", reversed_line)
